File: node-consumer/consumer.js
const { Kafka } = require("kafkajs");
const redis = require("redis");

const kafka = new Kafka({
  clientId: "node-consumer",
  brokers: [process.env.KAFKA_BOOTSTRAP_SERVERS || "kafka:9092"],
});

const topic = "bar-topic";
const redisHost = process.env.REDIS_HOST || "redis";
const redisClient = redis.createClient({ host: redisHost });

const consumer = kafka.consumer({ groupId: "node-consumer-group" });

const run = async () => {
  await consumer.connect();
  await consumer.subscribe({ topic, fromBeginning: true });

  await consumer.run({
    eachMessage: async ({ message }) => {
      const msg = message.value.toString();
      console.log(`Consumed message: ${msg}`);
      redisClient.incr("bar-counter", (err, reply) => {
        if (err) console.error(err);
        else console.log(`Updated Redis counter: ${reply}`);
      });
    },
  });
};

run().catch(console.error);

File: python-consumer/consumer.py
from kafka import KafkaConsumer
import redis
import os

KAFKA_BOOTSTRAP_SERVERS = os.environ.get(
    'KAFKA_BOOTSTRAP_SERVERS', 'kafka:9092')
TOPIC_NAME = 'foo-topic'
REDIS_HOST = os.environ.get('REDIS_HOST', 'redis')

consumer = KafkaConsumer(
    TOPIC_NAME,
    bootstrap_servers=KAFKA_BOOTSTRAP_SERVERS,
    auto_offset_reset='earliest',
    group_id='python-consumer-group'
)

r = redis.Redis(host=REDIS_HOST, port=6379, db=0)

for msg in consumer:
    print(f"Consumed message: {msg.value.decode('utf-8')}")
    r.incr('foo-counter')
    print(f"Updated Redis counter: {r.get('foo-counter').decode('utf-8')}")
